mava/utils/sebulba.py

```diff
@@ -197,3 +203,357 @@ def __enter__(self) -> None:
+
+        self.tickets_queue: queue.Queue = queue.Queue()
+        self._queue: queue.Queue = queue.Queue()
+        self.lifetime = lifetime
+
+        self.num_buffers = len(config.arch.actor_device_ids) * config.arch.n_threads_per_executor
+        self.rate_limiter = rate_limiter
+        self.sharding = learner_sharding
+        self.key = key
+
+        assert config.system.sample_batch_size % self.num_buffers == 0, (
+            f"The sample batch size ({config.system.sample_batch_size}) must be divisible "
+            f"by the total number of actors ({self.num_buffers})."
+        )
+
+        # Setup Buffers
+        rb = make_trajectory_buffer(
+            sample_sequence_length=config.system.sample_sequence_length + 1,
+            period=1,
+            add_batch_size=config.arch.num_envs,
+            sample_batch_size=config.system.sample_batch_size // self.num_buffers,
+            max_length_time_axis=config.system.buffer_size,
+            min_length_time_axis=config.system.min_buffer_size,
+        )
+        self.buffer_states = [rb.init(init_transition) for _ in range(self.num_buffers)]
+        self.buffer_adds_count = [0] * self.num_buffers
+
+        # Setup functions
+        self.buffer_add = jax.jit(rb.add, device=self.cpu)
+        self.buffer_sample = jax.jit(rb.sample, device=self.cpu)
+        self.move_to_device = lambda tree: jax.tree.map(lambda x: jax.device_put(x, self.cpu), tree)
+
+    def run(self) -> None:
+        """This function ensures that trajectories on the queue are consumed in the right order. The
+        start_condition and end_condition are used to ensure that only 1 thread is processing an
+        item from the queue at one time, ensuring predictable memory usage.
+        """
+        while not self.lifetime.should_stop():
+            try:
+                start_condition, end_condition = self.tickets_queue.get(timeout=1)
+                with end_condition:
+                    with start_condition:
+                        start_condition.notify()
+                    end_condition.wait()
+            except queue.Empty:
+                continue
+
+    def put(self, traj: Sequence[Transition], metrics: Tuple, actor_id: int) -> None:
+        start_condition, end_condition = (threading.Condition(), threading.Condition())
+        with start_condition:
+            self.tickets_queue.put((start_condition, end_condition))
+            start_condition.wait()
+
+        try:
+            self.rate_limiter.await_can_insert(timeout=QUEUE_PUT_TIMEOUT)
+        except TimeoutError:
+            print(
+                f"{Fore.RED}{Style.BRIGHT}Actor has timed out on insertion, "
+                f"this should not happen. A deadlock might be occurring{Style.RESET_ALL}"
+            )
+
+        # [Transition(num_envs)] * rollout_len -> Transition[done=(num_envs, rollout_len, ...)]
+        traj = _stack_trajectory(traj)
+        traj = jax.device_get(traj)
```

Comment:
Maybe be better to do the device get first so we block the device as little as possible? Although it's probably a tiny difference

```suggestion
        traj = jax.device_get(traj)
        traj = _stack_trajectory(traj)
```